Routes+validation

brandonbeckwith-ccm opened the pull request.

brandonbeckwith-ccm committed:
Added edit page, refactored http requests and added basic validation

brandonbeckwith-ccm committed:
Restructured / Modularized + finished Sean's assignment

Did a lot of CSS practice along with working with components. I mostly worked on breaking down elements / features into a more 're-usable' format.

brandonbeckwith-ccm committed:
Trying to get gh pages working...

brandonbeckwith-ccm committed:
Initial dist subtree commit

brandonbeckwith-ccm committed:
Updated gitignore and gh-pages

brandonbeckwith-ccm committed:
Updated README to include a link to the deployed version.

ethan-ccm commented on src/util/HttpRequests.ts:
Typescript files should start with a lowercase letter

ethan-ccm commented on src/components/AddTodo.vue:
```diff
@@ -0,0 +1,92 @@
+
+const addInput = defineModel<string>()!;
+
+const emits = defineEmits<{ (e: "submit", invalid: boolean): void }>();
```
Convention is to name this variable `emit` singular. The return type of `defineProps` is plural because it contains all props but here the return type isn't an object or array of all emits, it's a function with which you do the emitting

ethan-ccm commented on src/components/FlexCard.vue:
I'd recommend using this div directly instead of creating a component. Creating components for structuring/styling is fine but it should probably at least take some props or abstract away a significant amount of html.

ethan-ccm commented on src/components/FlexList.vue:
Same as above

ethan-ccm commented on src/components/IconButton.vue:
```diff
@@ -0,0 +1,45 @@
+  >
+    <slot></slot>
+  </button>
+</template>
```
Attribute inheritance is the default so you should be able to remove the explicit v-bind, as well as the inheritAttrs: false

ethan-ccm commented on src/components/TableItem.vue:
```diff
@@ -1,53 +1,127 @@
 const emit = defineEmits<{
-  (e: Events, taskInfo: TaskInfo): void
-}>()
+  (e: Events, taskInfo: TaskInfo): void;
```
This is an interesting approach but I'm not sure I've ever seen defineEmits used this way with a string union. If you, for some reason, had a lot of components with the same exact emits in them maybe this would be a clever way to reduce code duplication, but in general it's probably better to define each emit separately

ethan-ccm commented on src/components/TableItem.vue:
```diff
@@ -1,53 +1,127 @@
-let textInput = ref(props.todoName)
-let checkBoxInput = ref(props.isComplete)
+let textInput = ref(props.todoName);
+let checkBoxInput = ref(props.isComplete);
```
You should always basically always use `const` unless you plan on reassigning the variable later

ethan-ccm commented on src/views/TodoView.vue:
```diff
@@ -0,0 +1,329 @@
+const filterInput = ref("");
+const addText = ref("");
+
+type SORT_OPTIONS = "task_asc" | "task_dsc" | "complete" | "incomplete";
+
+const sortingOptions: { text: string; value: SORT_OPTIONS }[] = [
+  { text: "Tasks (dsc)", value: "task_dsc" },
+  { text: "Tasks (asc)", value: "task_asc" },
+  { text: "Complete", value: "complete" },
+  { text: "Incomplete", value: "incomplete" },
+];
```
Alternatively...
```typescript
const sortingOptions = [
  { text: "Tasks (dsc)", value: "task_dsc" },
  { text: "Tasks (asc)", value: "task_asc" },
  { text: "Complete", value: "complete" },
  { text: "Incomplete", value: "incomplete" },
] as const;
type SORT_OPTIONS = (typeof sortingOptions)[number]['value']
```
In your code the type is the source of truth and there will be an error if your actual options deviate from the type. In the alternative the options themselves are the source of truth and the type will automatically update whenever new options are added. Which is better depends on the situation.

Also types should ideally be formatted like you would a class name (i.e. SortOptions)

brandonbeckwith-ccm replied:
I can see why using `['value']` extracts the values, but why does `[number]` convert it to a union type?

ethan-ccm replied:
[This should explain things](https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html)

ethan-ccm commented on src/views/TodoView.vue:
```diff
@@ -0,0 +1,329 @@
+  { text: "Incomplete", value: "incomplete" },
+];
+
+const todoFiltered = computed(() => {
+  return [...todoData.value].filter((value) => {
+    // Copy or we end up blowing the stack
+    return value.todoName
+      .toLowerCase()
+      .includes(filterInput.value.toLowerCase());
+  });
+});
```
Can be simplified by removing braces and return statements, and I'm not sure why you're using spread here

brandonbeckwith-ccm replied:
I used spread to copy the values because for some reason there was an odd interaction with the ref causing endless recursive calls.

ethan-ccm replied:
I really am not sure why that would be happening I'd love if you could show me that

brandonbeckwith-ccm replied:
Weird, it's no longer an issue. I was probably updating the ref this uses somewhere causing endless calls...

ethan-ccm commented on src/views/TodoView.vue:
```diff
@@ -0,0 +1,329 @@
+  };
+
+  const completeSort = (a: ResponseData, b: ResponseData) => {
+    return ~~b.isComplete - ~~a.isComplete;
```
Had to google this one... This is cute, but in code that other people have to read I would just use the Number constructor to be explicit about your intentions. And here's some other ways to convert booleans to numbers https://stackoverflow.com/a/66000126. If you were gonna pick a shorthand to avoid typing out Number I'd probably go with the unary plus operator or a ternary as it's a little more obvious

ethan-ccm replied:
`!!` is very common though so feel free to use that instead of the `Boolean` constructor

brandonbeckwith-ccm replied:
Yeah I like the unary plus, ternary seems a bit overkill. I feel that the `+` is used commonly enough to convert to numbers that it should be readible enough.

ethan-ccm commented on src/views/TodoView.vue:
```diff
@@ -0,0 +1,329 @@
+    return ~~b.isComplete - ~~a.isComplete;
+  };
+
+  // Required so vue updates.
+  // Sort / reverse modify the original array
+  // which doesn't register as a change!
+  const sorting = [...todoFiltered.value];
```
`toSorted` fixes this issue but you might have to modify your ts config to target a higher JS version for it to recognize it as a valid method

brandonbeckwith-ccm committed:
Worked on putting the review to use.

Not everything was updated / adjusted due to breakage.

Fixed a bug with duplicate posts being sent.

brandonbeckwith-ccm committed:
Remove needless copying

brandonbeckwith-ccm committed:
Deployment file sync

brandonbeckwith-ccm committed:
Fixed name
